### PR TITLE
fix: cache warmup solution non legacy charts.

### LINKED ISSUE #15497: cache key for warm_up_cache and explore_json are not matching
Both the /warm_up_cache/ and /explore_json/ API endpoints should produce the same cache key for a given slice

### Expected results

should return same cache key 

### Actual results

different cache key are getting generated

#### How to reproduce the bug

1. run cache warmup for a slice
2. check the cache key value
3. explore the same slice in browser
4. different cache key will be generated so query will be fired.

### Environment

(please complete the following information):

- superset version: `1.1.0`
- python version: `3.8.9`
- node.js version: `12`

### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [ ] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [x] I have reproduced the issue with at least the latest released version of superset.
- [x] I have checked the issue tracker for the same issue and I haven't found one similar.

### Additional context

**Json_Data for Explore chart**

{"changed_on":1621922272000.0,"columns":["app_name"],"datasource":"1__table","extra_cache_keys":[],"extras":{"having":"","having_druid":[],"time_grain_sqla":"P1D","time_range_endpoints":["inclusive","exclusive"]},"filter":[],"granularity":null,"groupby":[],"is_rowcount":false,"is_timeseries":false,"metrics":[{"aggregate":"COUNT_DISTINCT","column":{"column_name":"app_id","description":null,"expression":null,"filterable":true,"groupby":true,"id":5,"is_dttm":false,"python_date_format":null,"type":"STRING","verbose_name":null},"expressionType":"SIMPLE","hasCustomLabel":false,"isNew":false,"label":"COUNT_DISTINCT(app_id)","optionName":"metric_6qp914zxuux_o2r0useg7ob","sqlExpression":null}],"order_desc":true,"orderby":[[{"aggregate":"COUNT_DISTINCT","column":{"column_name":"app_id","description":null,"expression":null,"filterable":true,"groupby":true,"id":5,"is_dttm":false,"python_date_format":null,"type":"STRING","verbose_name":null},"expressionType":"SIMPLE","hasCustomLabel":false,"isNew":false,"label":"COUNT_DISTINCT(app_id)","optionName":"metric_6qp914zxuux_o2r0useg7ob","sqlExpression":null},false]],"rls":[],"row_limit":100,"row_offset":0,"time_range":"Last week","timeseries_limit":0,"timeseries_limit_metric":null}

**Json Data for warmup** 

{"changed_on":"2021-05-25T05:57:52","datasource":"1__table","extra_cache_keys":[],"extras":{"druid_time_origin":"","having":"","having_druid":[],"time_grain_sqla":"P1D","time_range_endpoints":["inclusive","exclusive"],"where":""},"filter":[],"granularity":null,"groupby":["app_name"],"is_timeseries":false,"metrics":[{"aggregate":"COUNT_DISTINCT","column":{"column_name":"app_id","description":null,"expression":null,"filterable":true,"groupby":true,"id":5,"is_dttm":false,"python_date_format":null,"type":"STRING","verbose_name":null},"expressionType":"SIMPLE","hasCustomLabel":false,"isNew":false,"label":"COUNT_DISTINCT(app_id)","optionName":"metric_6qp914zxuux_o2r0useg7ob","sqlExpression":null}],"order_desc":true,"orderby":[[{"aggregate":"COUNT_DISTINCT","column":{"column_name":"app_id","description":null,"expression":null,"filterable":true,"groupby":true,"id":5,"is_dttm":false,"python_date_format":null,"type":"STRING","verbose_name":null},"expressionType":"SIMPLE","hasCustomLabel":false,"isNew":false,"label":"COUNT_DISTINCT(app_id)","optionName":"metric_6qp914zxuux_o2r0useg7ob","sqlExpression":null},false]],"rls":[],"row_limit":100,"time_range":"Last week","timeseries_limit":0,"timeseries_limit_metric":null}

### PR DESCRIPTION
**Pull request by dheeraj-jaiswal-lowes**
### SUMMARY
**Issue:** Cache-warmup API does not work as expected for all charts. Running cache-warmup for a slice ex: (vizType-Table) and exploring the same slice on browser gives different cache key.

**Solution:** /warm_up_cache/ API in current solution only applicable for legacy chart and those are very few in numbers, and when the same code executes for another type of chart such as Table chart and other non-legacy chart, it generates a different cache key not matching with the one we get while exploring the same slice on browser. Fix given here will check the viz type of the chart and handle it in the required way to update the right cache-key.

### TESTING INSTRUCTIONS
1. Create a dashboard that includes table chart.
2. Configure cache warm up schedule in superset_config.py.
3. Verify if the cache got refreshed after cache warm up execution using menu option at top-right corner of the chart.

### ADDITIONAL INFORMATION

- [x] Has associated issue: Fixes #15497
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/constants.py [hunk 1]
```diff
@@ -189,3 +189,32 @@ class CacheRegion(str, Enum):
...
+    "compare",
+    "partition",
+    "event_flow",
+    "world_map",
+    "paired_ttest",
+    "para",
+    "bullet",
+    "chord",
+    "pivot_table",
+]
```
Comment (villebro) [inline superset/constants.py:220]: I think a more dynamic approach would be to check if there's a query context for a given slice. If there is, assume it's the new type, and if there isn't, assume it's legacy.

Comment (dheeraj-jaiswal-lowes) [reply superset/constants.py:220]: @villebro Right. But there are scenarios where query context are not found even for new charts. Like When you Import a chart query context is not found. it is handled in Alert and reports using a function called "update_query_context". I found no other way to differentiate between charts. Please suggest if you have any other idea.

**Review by villebro — APPROVED** [review]
LGTM (barring any CI issues), great work!

**Comment by dheeraj-jaiswal-lowes** [comment]
New charts needs query-context saved to make cache-warmup work as expected. As per current behaviour, query-context are not included in chart's metadata when you export a chart. Above commit is to include the query-context when a chart is exported/imported.

**Comment by john-bodley** [comment]
@villebro per your comment, 

> I think a more dynamic approach would be to check if there's a query context for a given slice. If there is, assume it's the new type, and if there isn't, assume it's legacy.

It seems that many legacy charts include the `query_context` and thus this change actually breaks things, i.e., from Airbnb's production data we see:

```
mysql> SELECT DISTINCT viz_type FROM slices WHERE query_context IS NOT NULL ORDER BY viz_type;
+----------------------------+
| viz_type                   |
+----------------------------+
| area                       |
| bar                        |
| big_number                 |
| big_number_total           |
| box_plot                   |
| bubble                     |
| cal_heatmap                |
| chord                      |
| compare                    |
| country_map                |
| deck_polygon               |
| deck_scatter               |
| dist_bar                   |
| dual_line                  |
| echarts_area               |
| echarts_timeseries         |
| echarts_timeseries_bar     |
| echarts_timeseries_line    |
| echarts_timeseries_scatter |
| echarts_timeseries_smooth  |
| echarts_timeseries_step    |
| event_flow                 |
| filter_box                 |
| funnel                     |
| gauge_chart                |
| graph_chart                |
| heatmap                    |
| histogram                  |
| horizon                    |
| line                       |
| mapbox                     |
| mixed_timeseries           |
| para                       |
| partition                  |
| pie                        |
| pivot_table                |
| pivot_table_v2             |
| radar                      |
| sankey                     |
| sunburst                   |
| table                      |
| time_table                 |
| treemap                    |
| treemap_v2                 |
| tree_chart                 |
| word_cloud                 |
| world_map                  |
+----------------------------+
47 rows in set (0.34 sec)
```

**Comment by villebro** [comment]
> @villebro per your comment,
> 
> > I think a more dynamic approach would be to check if there's a query context for a given slice. If there is, assume it's the new type, and if there isn't, assume it's legacy.
> 
> It seems that many legacy charts include the `query_context` and thus this change actually breaks things, i.e., from Airbnb's production data we see:
> 
> ```
> mysql> SELECT DISTINCT viz_type FROM slices WHERE query_context IS NOT NULL ORDER BY viz_type;
> +----------------------------+
> | viz_type                   |
> +----------------------------+
> | area                       |
> | bar                        |
> | big_number                 |
> | big_number_total           |
> | box_plot                   |
> | bubble                     |
> | cal_heatmap                |
> | chord                      |
> | compare                    |
> | country_map                |
> | deck_polygon               |
> | deck_scatter               |
> | dist_bar                   |
> | dual_line                  |
> | echarts_area               |
> | echarts_timeseries         |
> | echarts_timeseries_bar     |
> | echarts_timeseries_line    |
> | echarts_timeseries_scatter |
> | echarts_timeseries_smooth  |
> | echarts_timeseries_step    |
> | event_flow                 |
> | filter_box                 |
> | funnel                     |
> | gauge_chart                |
> | graph_chart                |
> | heatmap                    |
> | histogram                  |
> | horizon                    |
> | line                       |
> | mapbox                     |
> | mixed_timeseries           |
> | para                       |
> | partition                  |
> | pie                        |
> | pivot_table                |
> | pivot_table_v2             |
> | radar                      |
> | sankey                     |
> | sunburst                   |
> | table                      |
> | time_table                 |
> | treemap                    |
> | treemap_v2                 |
> | tree_chart                 |
> | word_cloud                 |
> | world_map                  |
> +----------------------------+
> 47 rows in set (0.34 sec)
> ```
> 
> Given this is a regression I think that in the interim this should be reverted. Moving forward I wonder if there's a better failsafe mechanism, i.e., maybe try the `ChartDataCommand` route and fallback to the legacy logic if there's an error.

This is curious - legacy charts really shouldn't have a `query_context`, as they don't use it for anything. I wonder if this is caused by first creating a non-legacy chart, running that, then changing chart types. If this is indeed so, we should make sure `query_context` is always set to `null` if the chart doesn't use it.

**Comment by villebro** [comment]
> Given this is a regression I think that in the interim this should be reverted. Moving forward I wonder if there's a better failsafe mechanism, i.e., maybe try the ChartDataCommand route and fallback to the legacy logic if there's an error.

@john-bodley I'm not sure reverting this is the correct course of action, as prior to this PR, it was in fact impossible to warm up non-legacy charts all together. To get the best of both worlds I'd recommend getting a hotfix in which is able to do fallback to legacy logic if there's an error, and then following up with a PR that makes sure legacy charts clear out the query context if it's set. I can look into the latter if needed.

**Comment by john-bodley** [comment]
@villebro I'm not sure your hypothesis is correct, though I don't have one myself, as we (Airbnb) see numerous filter-box charts with the `query_context` defined and I speculate that the authors never changed the chart type, i.e., it's very atypical for someone to switch to a filter-box from another visualization type.

I also acknowledge that this PR tried to remedy a known issue, i.e., non-legacy charts not being cached, but in doing so it actually introduced two regressions:

1. Previous legacy charts were correctly being cached now fail.
2. Per https://github.com/apache/superset/pull/23569 it actually resulted in errors being thrown.

For these type of scenarios I sense we should err on the side of preventing regressions and thus revert said logic ASAP—hoping it hasn't been released into the wild in terms of a PyPI release—before formulating a correct implementation.

**Pull request merged**